### PR TITLE
rpc: rework binary encoding.  BREAKING CHANGE

### PR DESCRIPTION
**Pull request by mvines**
#### Problems:
1. base58 is slow, not all RPC method have a base64 encoding option
2. The usage of term "binary" for base58 is confusing
3. The usage of term "binary64" for base64 makes it worse
4. UiAccountData is not deserializable from Rust due to the untagged `Binary(String)` and `Binary64(String)` enums, requiring RpcClient hacks
5. RPC _binary_ results that don't include the encoding type will make it too difficult add new binary encodings in the future without breaking changes 

#### Changes:
1. Add base64 encoding support to getConfirmedTransaction/getConfirmedBlock
2. "binary" encoding is no longer documented in the JSON RPC API (but still supported).  A new "base58" encoding is added
3. "binary64" encoding is renamed to "base64".  **<-- BREAKING CHANGE**
4. RPC endpoints that return a `UiAccountData` for the "base58"/"base64" encoding now return an array of two elements, `["encoded data", "encoding type"]` instead of just `"encoded data"`.   This allows serde_json to work.  **<-- BREAKING CHANGE**
5. The binary format change to resolve 4 solves this problem as well.

#### TODO
* [ ] Inform downstream users of binary64 that they need to adjust their RPC calls, or update web3.js

### CODE REVIEW

File: transaction-status/src/lib.rs [hunk 1]
```diff
@@ -218,14 +218,16 @@ pub struct TransactionWithStatusMeta {
...
     Json,
     JsonParsed,
 }
 
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase", untagged)]
 pub enum EncodedTransaction {
-    Binary(String),
+    Binary(String),   // SLOW! Avoid this encoding
+    Binary64(String), // base-64 encoded transaction
```
Comment (mvines) [inline transaction-status/src/lib.rs:230]: 😭 serde_json can't distinguish between these two types during deserialization.

Comment (mvines) [reply transaction-status/src/lib.rs]: Cleaned this up and the fixed up the `UiAccountData` hacks along the way

File: account-decoder/src/lib.rs [hunk 2]
```diff
@@ -32,17 +32,18 @@ pub struct UiAccount {
...
     Json(ParsedAccount),
-    Binary64(String),
+    Binary(String, UiAccountEncoding),
 }
 
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
 #[serde(rename_all = "camelCase")]
 pub enum UiAccountEncoding {
-    Binary, // SLOW! Avoid this encoding
+    Binary, // Legacy. Retained for RPC backwards compatibility
```
Comment (mvines) [inline account-decoder/src/lib.rs:43]: One day we can remove "binary" when it starts getting in the way.  It's no longer documented

File: client/src/rpc_client.rs [hunk 3]
```diff
@@ -487,17 +482,8 @@ impl RpcClient {
...
                 } = serde_json::from_value::<Response<Option<UiAccount>>>(result_json)?;
-                if let Some(ref mut account) = rpc_account {
-                    if let Binary(_) = &account.data {
-                        let tmp = Binary64(String::new());
-                        match std::mem::replace(&mut account.data, tmp) {
-                            Binary(new_data) => account.data = Binary64(new_data),
-                            _ => panic!("should have gotten binary here."),
-                        }
-                    }
-                }
```
Comment (mvines) [inline client/src/rpc_client.rs:500]: The problem with `Binary(String)` and `Binary64(String)` is visible here.  Serde can't distinguish between the two on deserialization with the the `untagged` attribute, so this workaround was previously needed for Rust RPC clients.

File: core/src/rpc.rs [hunk 4]
```diff
@@ -3119,24 +3122,27 @@ pub mod tests {
         bank.store_account(&address, &account);
 
         let req = format!(
-            r#"{{"jsonrpc":"2.0","id":1,"method":"getAccountInfo","params":["{}", {{"encoding":"binary64"}}]}}"#,
+            r#"{{"jsonrpc":"2.0","id":1,"method":"getAccountInfo","params":["{}", {{"encoding":"base64"}}]}}"#,
```
Comment (mvines) [inline core/src/rpc.rs:3125]: Ah, this feels so much nicer.  "binary64"?  What's that!  "base64"?  Oh, that's base64.

**Comment by mvines** [comment]
This PR escalated quickly!   It's a breaking change mostly for the newly added binary64 encoding, so the downstream users should still be few and I think the long-term benefits outweigh the minor inconvenience of adapting to the slightly different RPC request/responses

File: core/src/rpc.rs (hunk 4 above)
Comment (jstarry) [reply core/src/rpc.rs:3125]: Yes yes yes!! 🙌

**Pull request merged**